course-content/31_exploring_data_structures.qmd

```diff
@@ -0,0 +1,298 @@
+:::
+
+:::{.callout-warning icon="false"}
+
+## Challenge 1
+
+
+
+Let's say you just want to narrow down the dataset to include just the country and the GDP per capita. How would you do it? Don't forget to run your code so it also shows a view of the result so you can confirm the code worked as you wanted it to. 
+
+:::
+
+
+
+:::{.callout-note icon="false" collapse="true"}
+
+### Solution to Challenge 1
+
+First you would like to create a new object to hold this narrowed down version of the dataset, then just use some method like `.head()` to see some of the resulting data. 
+
+```{python}
+challenge1_df = df[['country', 'gdpPercap']]
+challenge1_df.head(2)
+```
+
+:::
+
+### Selecting Rows
+
+Of course, in data analysis we are usually interested in looking at just some of rows of data, not all of it all the time. So when we want to look at just selected rows (i.e. values within a given column) there is a handy pandas function called `.query()` that we can leverage. It helps us with selecting the rows we want. 
+
+We set up an expression in pandas  with `.query()` and it will evaluate it for us. For example, imagine we want to find just the rows where the year is 1972. Here's how we do that, using `==` instead of just `=`:
+
+```{python}
+filtered_df = df.query('year == 1972')
+filtered_df.head(2)
+```
+
+We can even get more specific with string variables, looking for content by starting letter or letters.  For example, here's how we can look for specific countries: 
+
+```{python}
+filtered_df = df.query('country.str.startswith("C") & year == 1972')
+filtered_df.head(3)
+```
+
+:::{.callout-warning icon="false"}
+
+## Challenge 2
+
+
+
+Your director has come to you and asked if you know what the life expectancy was is in Canada in 2007.  How would you use pandas code to get the data you need? And after having run the code, what's the answer? 
+
+:::
+
+
+
+:::{.callout-note icon="false" collapse="true"}
+
+### Solution to Challenge 2
+
+First you would like to create a new object to hold this narrowed down version of the dataset, then just use some method like `.print()` or `.head()` to see some of the resulting data. 
+
+```{python}
+challenge2_df = df.query('country.str.startswith("Canada") & year == 1972')
+print(challenge2_df)
+```
+
+:::
+
+:::{.callout-tip}
+## .loc() and .iloc() to also filter rows! 
+There is something else you should know about row selection. It is VERY common within python/pandas to see the `.loc()` and `.iloc()` methods to identify rows by index location. The index starts at 0 for the first row (yes, "0", and not "1")!  You can also use these methods. You may not ever need to use it, but it is good to know what it means when you come across it. 
```

Comment:
If you refer to `.iloc()` you should also have an example, and talk about when you would use each... otherwise I think you should only mention one.